tests/end_to_end/end_to_end.cpp

```diff
@@ -386,13 +386,226 @@ bindmonitor_test(ebpf_execution_type_t execution_type)
+    return result;
+}
+
+void
+countbytes_test(ebpf_execution_type_t execution_type)
+{
+    _test_helper_end_to_end test_helper;
+
+    const char* error_message = nullptr;
+    bpf_object* object = nullptr;
+    bpf_link* link = nullptr;
+    fd_t program_fd = 0ul;
+    ebpf_result_t result{};
+
+    program_info_provider_t mac_program_info(EBPF_PROGRAM_TYPE_MAC);
+
+    result = ebpf_program_load(
+        SAMPLE_PATH "countbytes.o", &EBPF_PROGRAM_TYPE_MAC, &EBPF_ATTACH_TYPE_MAC, execution_type, &object, &program_fd, &error_message);
+    if (error_message) {
+        printf("ebpf_program_load failed with %s\n", error_message);
+        ebpf_free_string(error_message);
+        error_message = nullptr;
+    }
+    REQUIRE(result == EBPF_SUCCESS);
+
+    fd_t byte_map_fd = bpf_object__find_map_fd_by_name(object, "byte_map");
+    REQUIRE(byte_map_fd > 0);
+
+    single_instance_hook_t hook(EBPF_PROGRAM_TYPE_MAC, EBPF_ATTACH_TYPE_MAC);
+
+    REQUIRE(hook.attach_link(program_fd, &link) == EBPF_SUCCESS);
+
+    // Send IPV4 UDP packet
+    uint64_t fake_length_1 = 100;
+    REQUIRE(emulate_mac(hook, fake_length_1, 0x11, 1) == 0);
+    REQUIRE(result == 0);
+
+    // Send IPV4 TCP packet
+    uint64_t fake_length_2 = 200;
+    REQUIRE(emulate_mac(hook, fake_length_2, 0x06, 1) == 0);
+    REQUIRE(result == 0);
+
+    // Send IPV6 UDP packet
+    uint64_t fake_length_3 = 300;
+    REQUIRE(emulate_mac(hook, fake_length_3, 0x11, 0) == 0);
+    REQUIRE(result == 0);
+
+    // Send IPV6 TCP packet
+    uint64_t fake_length_4 = 400;
+    REQUIRE(emulate_mac(hook, fake_length_4, 0x06, 0) == 0);
+    REQUIRE(result == 0);
+
+    five_tuple_t key{};
+    uint64_t value{};
+
+    REQUIRE(bpf_map_get_next_key(byte_map_fd, NULL, &key) == EBPF_SUCCESS);
+    REQUIRE(key.protocol != 0);
+    REQUIRE(key.dest_port == 0);
+    REQUIRE(key.source_port == 0);
+    REQUIRE(bpf_map_lookup_elem(byte_map_fd, &key, &value) == EBPF_SUCCESS);
+    REQUIRE(value != 0);
+    REQUIRE(bpf_map_get_next_key(byte_map_fd, &key, &key) == EBPF_SUCCESS);
+    REQUIRE(key.protocol != 0);
+    REQUIRE(key.dest_port == 0);
+    REQUIRE(key.source_port == 0);
+    REQUIRE(bpf_map_lookup_elem(byte_map_fd, &key, &value) == EBPF_SUCCESS);
+    REQUIRE(value != 0);
+    REQUIRE(bpf_map_get_next_key(byte_map_fd, &key, &key) == EBPF_SUCCESS);
+    REQUIRE(key.protocol != 0);
+    REQUIRE(key.dest_port == 0);
+    REQUIRE(key.source_port == 0);
+    REQUIRE(bpf_map_lookup_elem(byte_map_fd, &key, &value) == EBPF_SUCCESS);
+    REQUIRE(value != 0);
+    REQUIRE(bpf_map_get_next_key(byte_map_fd, &key, &key) == EBPF_SUCCESS);
+    REQUIRE(key.protocol != 0);
+    REQUIRE(key.dest_port == 0);
+    REQUIRE(key.source_port == 0);
+    REQUIRE(bpf_map_lookup_elem(byte_map_fd, &key, &value) == EBPF_SUCCESS);
+    REQUIRE(value != 0);
+    REQUIRE(bpf_map_get_next_key(byte_map_fd, &key, &key) == -EBPF_NO_MORE_KEYS);
+
+    hook.detach_link(link);
```

Comment:
This cleanup should be done in smart objects in case any exceptions are thrown (here and above)